src/avisi/atlassian/connect/firestore.clj

```diff
@@ -0,0 +1,195 @@
+  [(s/keys :req [::host-client-key ::document-path]) => (? map?)]
+  (fetch-document {::document-path (format "hosts/%s/%s" host-client-key document-path)}))
+
+(>defn get-host-sub-collection-documents
+  [{::keys [host-client-key collection-path]}]
+  [(s/keys :req [::host-client-key ::collection-path]) => vector?]
+  (let [doc-refs (->
+                   ^Firestore service
+                   ^CollectionReference (.collection (format "hosts/%s/%s" host-client-key collection-path))
+                   ^"[Lcom.google.cloud.firestore.DocumentReference;" (.listDocuments))]
+    (mapv doc-ref->entity doc-refs)))
+
+(defn find-document [{::keys [collection field value]}]
+  (->
+    ^Firestore service
+    ^Query (.collectionGroup collection)
+    ^Query (.whereEqualTo ^FieldPath (FieldPath/of (into-array [field])) value)
+    ^ApiFuture (.get)
+    ^QuerySnapshot (.get)
+    ^"[Lcom.google.cloud.firestore.QueryDocumentSnapshot;" (.getDocuments)
+    ^QueryDocumentSnapshot first
+    ^DocumentReference (.getReference)))
+
+(defn find-entity [env]
+  (->
+    (find-document env)
+    (doc-ref->entity)))
+
+(defn get-document-parent-entity [env]
+  (->
+    ^DocumentReference (find-document env)
+    ^CollectionReference (.getParent)
+    ^DocumentReference (.getParent)
+    doc-ref->entity))
+
+(defn delete-document [^DocumentReference doc-ref] (.get (.delete doc-ref)))
+
+(>defn get-host
+  [host-client-key]
+  [string? => (? map?)]
+  (fetch-document {::document-path (format "hosts/%s" host-client-key)}))
+
+(defn collection-entities [collection]
+  (mapv doc-ref->entity
+    (->
+      ^Firestore service
+      ^CollectionReference (.collection "hosts")
+      ^"[Lcom.google.cloud.firestore.DocumentReference;" (.listDocuments))))
+
+(defn namespace-all-keys
+  "Change all keys from camelCase to kebab-case namespace keywords
+  Example: clientKey will become :atlassian-connect.host/client-key"
+  [m]
+  (reduce (fn [m [k v]] (assoc m (keyword "atlassian-connect.host" (csk/->kebab-case (name k))) v)) {} m))
+
+(defn update-host!
+  [{:keys [prev-host new-host]
+    :as env}]
+  (store-host! {::document (merge prev-host new-host)}))
+
+(defn handle-lifecycle-payload! [request]
+  (let [lifecycle-payload (namespace-all-keys (:body-params request))
+        _ (log/info
+            (select-keys
+              lifecycle-payload
+              [:atlassian-connect.host/base-url :atlassian-connect.host/client-key :atlassian-connect.host/event-type])
+            "Handling host lifecycle call")
+        client-key (host-client-key lifecycle-payload)
+        prev-host (get-host client-key)]
+    ;; Check JWT token on subsequent calls, this prevents tinkering with shared-secrets of hosts.
+    (when prev-host
+      (jwt/validate-jwt-token
+        request
+        {:validate-qsh? true
+         :shared-secret (:atlassian-connect.host/shared-secret prev-host)}))
+    (update-host!
+      {:prev-host prev-host
+       :new-host lifecycle-payload})))
```

Comment:
Ik zou hier de payload omzetten naar hetzelfde format wat er nu is in de numbered-headings-cloud en edit-custom-field-values-cloud